File: examples/python/02_tools_structured_output.py
"""Tools plus typed, validated output."""

import asyncio
from dataclasses import dataclass
from typing import Literal

from openharness import Agent, run, tool


@tool
def get_weather(city: str) -> dict:
    """Get the current weather for a city.

    Args:
        city: City name, e.g. "Toronto".
    """
    fake = {"toronto": (14, "cloudy"), "paris": (21, "sunny"), "london": (11, "rain")}
    temp, cond = fake.get(city.lower(), (18, "cloudy"))
    return {"city": city, "temp_c": temp, "conditions": cond}


@dataclass
class TripAdvice:
    city: str
    temp_c: float
    conditions: Literal["sunny", "cloudy", "rain"]
    pack: list[str]


agent = Agent(
    name="Travel helper",
    instructions="Check the weather with your tool, then suggest what to pack.",
    tools=[get_weather],
    output_type=TripAdvice,
)


async def main() -> None:
    result = await run(agent, "I'm flying to London tomorrow.")
    advice: TripAdvice = result.output
    print(advice)
    print(f"{result.turns} turns, {result.usage.total_tokens} tokens")


asyncio.run(main())

File: examples/typescript/02_tools_structured_output.ts
// Tools plus typed, validated output.
import { Agent, run, s, tool, type Infer } from "openharness";

const getWeather = tool({
  name: "get_weather",
  description: "Get the current weather for a city.",
  parameters: s.object({ city: s.string().describe('City name, e.g. "Toronto".') }),
  execute: ({ city }) => {
    const fake: Record<string, [number, string]> = { toronto: [14, "cloudy"], paris: [21, "sunny"], london: [11, "rain"] };
    const [tempC, conditions] = fake[city.toLowerCase()] ?? [18, "cloudy"];
    return { city, tempC, conditions };
  },
});

const TripAdvice = s.object({
  city: s.string(),
  tempC: s.number(),
  conditions: s.enum(["sunny", "cloudy", "rain"]),
  pack: s.array(s.string()),
});
type TripAdvice = Infer<typeof TripAdvice>;

const agent = new Agent<TripAdvice>({
  name: "Travel helper",
  instructions: "Check the weather with your tool, then suggest what to pack.",
  tools: [getWeather],
  outputType: TripAdvice,
});

const result = await run(agent, "I'm flying to London tomorrow.");
const advice: TripAdvice = result.output;
console.log(advice);
console.log(`${result.turns} turns, ${result.usage.totalTokens} tokens`);
